web-api/src/storage/postgres.rs

```diff
@@ -493,72 +555,235 @@ impl Database {
-            Vec::new()
+        let property = if let Some(property) = property {
+            Some(Some(property.0 .0))
         } else {
-            let unmigrated = self.elastic.get_unmigrated(&ids).await?;
-            QueryBuilder::new(
-                "INSERT INTO document (document_id, snippet, properties, tags, embedding) ",
-            )
-            .push_values(&unmigrated, |mut builder, document| {
-                builder
-                    .push_bind(&document.id)
-                    .push_bind(&document.snippet)
-                    .push_bind(Json(&document.properties))
-                    .push_bind(&document.tags)
-                    .push_bind(&document.embedding);
-            })
-            .push(
-                " ON CONFLICT (document_id) DO UPDATE SET
-                snippet = EXCLUDED.snippet,
-                properties = EXCLUDED.properties,
-                tags = EXCLUDED.tags,
-                embedding = EXCLUDED.embedding
-                RETURNING document_id;",
-            )
-            .build()
-            .persistent(false)
-            .try_map(|row| DocumentId::from_row(&row))
-            .fetch_all(&mut tx)
-            .await?
+            Database::document_exists(&mut tx, document_id)
+                .await?
+                .then_some(None)
         };
 
         tx.commit().await?;
 
-        if migrated.len() < ids.len() {
-            error!(expected = ?ids, actual = ?migrated, "Failed to migrate some documents");
-            Err(DocumentNotFound.into())
+        Ok(property)
+    }
+
+    async fn put(
+        &self,
+        document_id: &DocumentId,
+        property_id: &DocumentPropertyId,
+        property: &DocumentProperty,
+    ) -> Result<Option<()>, Error> {
+        let mut tx = self.postgres.pool.begin().await?;
+
+        let inserted = sqlx::query(
+            "UPDATE document
+            SET properties = jsonb_set(properties, $1, $2)
+            WHERE document_id = (
+                SELECT document_id
+                FROM document
+                WHERE document_id = $3
+                FOR UPDATE
+            );",
+        )
+        .bind(slice::from_ref(property_id))
+        .bind(Json(property))
+        .bind(document_id)
+        .execute(&mut tx)
+        .await?
+        .rows_affected();
+        let inserted = if inserted > 0 {
+            self.elastic
+                .insert_document_property(document_id, property_id, property)
+                .await?
         } else {
-            debug!(?ids, "Migrated {} documents", ids.len());
-            Ok(all_migrated)
-        }
+            None
+        };
+
+        tx.commit().await?;
+
+        Ok(inserted)
+    }
+
+    async fn delete(
+        &self,
+        document_id: &DocumentId,
+        property_id: &DocumentPropertyId,
+    ) -> Result<Option<Option<()>>, Error> {
+        let mut tx = self.postgres.pool.begin().await?;
+
+        let deleted = sqlx::query(
```

Comment:
:pushpin: I have been wondering if delete shouldn't be more idempotent, i.e. deleting a property which doesn't exist should IMHO not fail (if the document doesn't exist it should).